### ISSUE TITLE
EmbeddedDocumentSerializer not allowing a field to be optional

### ISSUE THREAD

**Issue by pkout** [umutbozkurt/django-rest-framework-mongoengine#172]
Hi. I have a serializer object like this:

```
from rest_framework import serializers
from rest_framework_mongoengine.serializers import DocumentSerializer
from rest_framework_mongoengine.serializers import EmbeddedDocumentSerializer

class IssueSerializer(EmbeddedDocumentSerializer):
    note = serializers.CharField(required=False, allow_blank=True)

    class Meta:
        model = Issue
        depth = 2

    def validate(self, data):
        return data

```

The document serializer that uses this embedded document is defined like this:

```
class ProjectSerializer(DocumentSerializer):
    events = IssueSerializer(Issue, many=True)

```

The embedded document model is defined like this:

```
class Issue(EmbeddedDocument):
    id = ObjectIdField(required=True, default=ObjectId())
    status = StringField(required=True, max_length=50)
    note = StringField(required=False, max_length=2000)
    created = DateTimeField(required=True, default=timezone.now())

```

And the Project document is defined like this:

```
class Project(Document):
    author = ReferenceField(Author, required=True)
    short_description = StringField(required=True, max_length=255)
    issues = EmbeddedDocumentListField(Issue, required=False)

```

As you see, the `note` field is defined as optional (required=False) in both the model and the serializer. Yet, when I try to serialize the `Project` document with some issues in it, it returns a validation error saying that the `note` field is required. Am I doing this wrong? Please, advice if I need to change something? Is this a bug? Thank you!

**Comment by qwiglydee** [comment]
do you use last release or git master head?

also, `allow_blank` is invalid kwarg. it should be `allow_null`

**Comment by qwiglydee** [comment]
you have some issues elsewhere in your code.
here is test case that works (with v3.3.0):
```
class Issue(EmbeddedDocument):
    id = fields.ObjectIdField(required=True, default=ObjectId())
    status = fields.StringField(required=True, max_length=50)
    note = fields.StringField(required=False, max_length=2000)
    created = fields.DateTimeField(required=True, default=timezone.now())


class Project(Document):
    events = fields.EmbeddedDocumentListField(Issue)


class IssueSerializer(EmbeddedDocumentSerializer):
    note = serializers.CharField(required=False, allow_blank=True)

    class Meta:
        model = Issue
        depth = 2

    def validate(self, data):
        return data


class ProjectSerializer(DocumentSerializer):
    class Meta:
        model = Project

    events = IssueSerializer(Issue, many=True)


class TestSerializer(TestCase):
    def test_serializing(self):
        now = timezone.now()
        i = Issue(status="new", created=now)
        p = Project.objects.create(events=[i])
        s = ProjectSerializer(instance=p)
        assert s.data == {'id': str(p.id), 'events': [{'id': str(i.id), 'status': "new", 'created': now.isoformat(), 'note': None, }]}

```

**Comment by pkout** [comment]
Hi, 

Thanks for the quick reply! I wasn't specific enough in my post. I apologize - I realize that I reported the wrong scenario. It works for me when I serialize a brand new Project with a new embedded Issue object as well. If I don't set the "note" field there, it works fine - the field isn't created in the DB. However, it returns a validation error when I later load the serialized Project object from the DB and add a new Issue to it. Can you please try something along these lines?

def test_serializing(self):
    # Create a project with one issue and test that it was properly serialized
    now = timezone.now()
    i = Issue(status="new", created=now)
    p = Project.objects.create(events=[i])
    s = ProjectSerializer(instance=p)
    assert s.data == {'id': str(p.id), 'events': [{'id': str(i.id), 'status': "new", 'created': now.isoformat(), 'note': None }]}
    # Load the created project, add an issue to it and test that the new issue was serialized
    p = Project.objects.all().first()
    ser = ProjectSerializer(instance=p)
    d = ser.data
    d.events.append({'status': 'In Progress'})
    ser2 = ProjectSerializer(instance=p, data=d)
    if not ser2.valid():
        print ser2.errors
    else:
        ser2.save()
     assert len(s.data.events) == 2

I noticed that it starts working if I first delete the note element from the deserialized Project's first Issue dictionary in d. The issue.note value is set to None there (from the deserialization). If I try to use that dictionary to add a new issue to it and serialize the modified dictionary back to the DB, it gives me the validation error saying that note field can't be null. It looks like the serializer validator doesn't accept dictionary elements that are set to None even for fields that are set to be optional. I hope my explanation isn't too difficult to follow.

**Comment by pkout** [comment]
Okay, I just realized that I am doing the whole thing in an unnecessarily complicated way. I don't need to modify the data dictionary of the entire existing Project object. I can use partial=True style update and submit only the newly added issue:

```
p = Project.objects.all().first()
ser = ProjectSerializer(instance=p)
new_issue = {'events': [{'status': 'In Progress'}]}
ser2 = ProjectSerializer(instance=p, data=new_issue, partial=True)
if ser2.is_valid():
    ser2.save
else:
    print ser2.errors
```

This works and is much more concise. It would still be nice if dictionary items set to None didn't trigger "required" validation error when the corresponding field is set to required=False, but the "cleaner" way to do it is to avoid None values in the first place. Thanks!

**Comment by qwiglydee** [comment]
I cannot get how you come to None triggering validation.
Can you modify testcase code to match your situation and make it fail?

**Comment by pkout** [comment]
I modified your test code to trigger the error when I posted my first response to you earlier today. ;) Please, let me know if something isn't clear in the code.

**Comment by qwiglydee** [comment]
The problem was that the Issue is serialized with note=None and it cannot be parsed back, because you do not gave allow_null=True for the field.

**Comment by qwiglydee** [comment]
In general, it seems like you need nested API, like /projects/<id>/issues/
(it is not supported out of the box and needs reimplemention of nested operations, maybe i'll implement such support in DRFM, because nested docs are important in mongo)

You can manipulate individual elements:
to create: `project.update(push__events=new_issue)`
to update: `Project.objects.filter(events__issue=issue_id).update(set__events__S=replacement)`
to delete: `project.update(pull__events={id: issue_id})`

Also, you can try contrib/patching from master branch:
to create: PATCH /projects/<p_id> [{ 'path': "/events", 'op': 'push', 'value': { new issue data }}]
to update: PATCH /projects/<p_id>/issues/<i_id> [{ 'path': "/events/S/subfield", 'op': 'set', 'value': something }]
to delete: PATCH /projects/<p_id> [{ 'path': "/events", 'op': 'pull', 'value': { 'id': issue_id }}]

to retrieve single issue, either python-level filtering required,
or dirty hack with `queryset._loaded_fields = QueryFieldList(['events.$'])`

**Comment by pkout** [comment]
Thanks! Okay, once I changed the note field argument from allow_blank=True to allow_null=True, serializing the None field starting working.  I misunderstood the DRF documentation when it said it recommends using allow_blank instead of allow_null. It makes perfect sense now. Thanks again! This closes the issue.

I also tried your suggestions regarding: `project.update(push__events=new_issue)` and `projects.update(pull__events={id: issue_id})` and didn't work on my end:

```
project = Project.objects.all().first()
new_issue = Issue(status='new')
print 'before:', len(project['events'])
project.update(push__events=new_issue)
print 'after:', len(project['events'])
```

This outputs:
```
before: 3
after: 3
```

For `project.update(pull__events={id: issue_id})`, with this code:

```
issue_id = Project.objects.all().first()['events'][0].id
project.update(pull__events={id: issue_id})
```

I get this exception:

`FieldDoesNotExist: The field '<built-in function id>' does not exist on the document 'Issue'`

My Issue embedded document looks like this:

```
class Issue(EmbeddedDocument):
    id = ObjectIdField(required=True, default=ObjectId())
    status = StringField(required=True, max_length=50)
    note = StringField(required=False, max_length=2000)
    created = DateTimeField(required=True, default=timezone.now())

```

Perhaps this is because `EmbeddedDocument` can't work with ID's the way `Document` does?

I don't really need this feature because there's another way to do it, but since you suggested it, I thought it would be good to give you a feedback on it.

The following is totally unrelated to this issue, so if you want to work on this together perhaps, I'll start a new issue: I needed a basic session-based user authentication and it didn't seem to be implemented in the DRFM, so I forked the repo and built it in by extracting relevant pieces of code from the mongoengine.django package version 0.9 that still had support for Django and coding them into DRFM. I am not sure that I didn't do something wrong, but I didn't see any code to implement authentication in DRFM. Please, let me know if you'd like to look at my code. It authenticates users properly for me now.

Thanks!
